services/device-registry-base/src/test/java/org/eclipse/hono/deviceregistry/service/device/AutoProvisionerTest.java

```diff
@@ -153,6 +153,29 @@ public void testAutoProvisioningIsSuccessfulForAlreadyPresentEdgeDevice(final Ve
                 }));
     }
 
+    /**
+     * Verifies that auto-provisioning fails if two different gateways try to provision the same device concurrently.
+     *
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    public void testAutoProvisioningFailsWhenTriggeredConcurrentlyByDifferentGatewaysForTheSameDevice(final VertxTestContext ctx) {
+        mockAssertRegistration(GATEWAY_ID, Collections.singletonList(GATEWAY_GROUP_ID), Collections.singletonList(RegistryManagementConstants.AUTHORITY_AUTO_PROVISIONING_ENABLED));
+        mockAssertRegistration(DEVICE_ID, true);
+        mockAddEdgeDevice(HttpURLConnection.HTTP_CONFLICT);
+
+        autoProvisioner.performAutoProvisioning(Constants.DEFAULT_TENANT, DEVICE_ID, "another-" + GATEWAY_ID, NEW_EDGE_DEVICE, span.context())
+                .onComplete(ctx.failing(throwable -> {
+                    ctx.verify(() -> {
+                        verify(deviceManagementService).createDevice(any(), any(), any(), any());
```

Comment:
it helps understanding the test better, if you are as specific as possible in the verification. In this case, I guess this would be
```
verify(deviceManagementService).createDevice(eq(Constants.DEFAULT_TENANT), eq(Optional.of(DEVICE_ID)), any(Device.class), any(Span.class));
```
or am I mistaken?

Reply:
No, this is correct. 

One could be more specific by adding an argument captor checking if the given device hast GATEWAY_ID in its "via" property. Not sure where to draw the line though regarding "_as specific as possible_".
Personally I it feels a bit odd to verify what I just wrote in the same test, but I am open to add that as well if it helps.

I also added the assertions regarding the status code.